src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs

```diff
@@ -132,10 +133,29 @@ public async Task<bool> PushMetadataAsync(CancellationToken cancellationToken)
             return builds.Select(t => new BuildRef(t.Key, t.Value)).ToImmutableList();
         }
 
-        private async Task<int?> GetBuildId(DependencyDetail dep, IMaestroApi client, CancellationToken cancellationToken)
+        private static async Task<int?> GetBuildId(DependencyDetail dep, IMaestroApi client, Dictionary<(string name, string version), int> assetCache, CancellationToken cancellationToken)
         {
+            if (assetCache.TryGetValue((dep.Name, dep.Version), out int value))
+            {
+                return value;
+            }
             var assets = await client.Assets.ListAssetsAsync(name: dep.Name, version: dep.Version, cancellationToken: cancellationToken);
-            return assets.OrderByDescending(a => a.Id).FirstOrDefault()?.BuildId;
+            var buildId = assets.OrderByDescending(a => a.Id).FirstOrDefault()?.BuildId;
+            if (!buildId.HasValue)
+            {
+                return null;
+            }
```

Comment:
Would it hurt to cache this too to simplify the code? And just have it be `Dictionary<,int?>`?

Reply:
Ah, we can't because we are going to look up the build...

---
In reply to: [271050299](https://github.com/dotnet/arcade-services/pull/296#discussion_r271050299) [](ancestors = 271050299)